core/src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexIT.java

```diff
@@ -196,4 +206,59 @@ public void testInvalidShardCountSettingsWithoutPrefix() throws Exception {
         }
     }
 
+    public void testCreateAndDeleteIndexConcurrently() throws InterruptedException {
+        createIndex("test");
+        final AtomicInteger indexDeleteAction = new AtomicInteger(0);
+        final CountDownLatch latch = new CountDownLatch(1);
+        int numDocs = randomIntBetween(1, 10);
+        for (int i = 0; i < numDocs; i++) {
+            client().prepareIndex("test", "test").setSource("index_version", indexDeleteAction.get()).get();
+        }
+        synchronized (indexDeleteAction) { // not necessarily needed here
+            indexDeleteAction.incrementAndGet();
+        }
+        client().admin().indices().prepareDelete("test").execute(new ActionListener<DeleteIndexResponse>() { // this happens async!!!
+                @Override
+                public void onResponse(DeleteIndexResponse deleteIndexResponse) {
+                    Thread thread = new Thread() {
+                     public void run() {
+                         try {
+                             client().prepareIndex("test", "test").setSource("index_version", indexDeleteAction.get()).get(); // recreate that index
+                             synchronized (indexDeleteAction) {
+                                 indexDeleteAction.incrementAndGet();
+                             }
+                             client().admin().indices().prepareDelete("test").get(); // from here on all docs with index_version == 0|1 must be gone!!!! only 2 are ok;
```

Comment:
all I am asserting in this test is that we never reuse an old shard etc. to your not being so suprised this test failed a lot for different reasons if you have a better idea go ahead.